protocols/kad/src/jobs.rs

```diff
@@ -193,36 +193,33 @@ impl PutRecordJob {
     /// to be run.
     pub fn poll<T>(&mut self, cx: &mut Context<'_>, store: &mut T, now: Instant) -> Poll<Record>
     where
-        for<'a> T: RecordStore<'a>,
+        T: RecordStore,
     {
         if self.inner.check_ready(cx, now) {
             let publish = self.next_publish.map_or(false, |t_pub| now >= t_pub);
-            let records = store
-                .records()
-                .filter_map(|r| {
-                    let is_publisher = r.publisher.as_ref() == Some(&self.local_id);
-                    if self.skipped.contains(&r.key) || (!publish && is_publisher) {
-                        None
-                    } else {
-                        let mut record = r.into_owned();
-                        if publish && is_publisher {
-                            record.expires = record
-                                .expires
-                                .or_else(|| self.record_ttl.map(|ttl| now + ttl));
-                        }
-                        Some(record)
+            #[allow(clippy::mutable_key_type)]
+            let mut skipped = Default::default();
+            std::mem::swap(&mut skipped, &mut self.skipped);
+            let record_ttl = self.record_ttl;
+            let local_id = self.local_id;
+            let records = Box::new(store.records().filter_map(move |r| {
```

Comment:
`store.records` now returns an owned `Iter`. With the `filter_map` right after we might be cloning many items in vain. Isn't that an issue?